File: azure_functions/agents/Twitter_Apify_TS/src/supabaseRetriever.ts
import { createClient, SupabaseClient } from '@supabase/supabase-js';

/**
 * Create and return Supabase client
 */
export function getSupabaseClient(): SupabaseClient {
  const url = process.env.SUPABASE_URL;
  const key = process.env.SUPABASE_SERVICE_ROLE_KEY;

  if (!url || !key) {
    throw new Error('SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY must be set');
  }

  return createClient(url, key);
}

/**
 * Loads JSON content from a specific item in the 'items' table in Supabase.
 */
export async function loadJsonFromSupabase(itemName: string): Promise<any | null> {
  try {
    const supabase = getSupabaseClient();
    const { data, error } = await supabase
      .from('items')
      .select('payload')
      .eq('name', itemName)
      .single();

    if (error) {
      console.error(`Error loading ${itemName} from Supabase:`, error);
      return null;
    }

    if (data && 'payload' in data) {
      return data.payload;
    }

    return null;
  } catch (error) {
    console.error(`Error loading ${itemName} from Supabase:`, error);
    return null;
  }
}

/**
 * Get list of Twitter usernames to monitor from Supabase.
 * Returns list of usernames (without @).
 */
export async function getMonitoredTwitterUsernames(): Promise<string[]> {
  const config = await loadJsonFromSupabase('twitter_config');
  if (!config) {
    return [];
  }

  const usernames: string[] = [];
  const monitoredUrls = config.monitored_urls || [];

  for (const entry of monitoredUrls) {
    const profileUrl = entry.profile_url || '';
    if (!profileUrl) {
      continue;
    }

    // Extract username from URL like "https://x.com/@Lica2216"
    // Handle both x.com and twitter.com
    try {
      // Get last part of URL
      let username = profileUrl.replace(/\/$/, '').split('/').pop();
      if (!username) {
        continue;
      }
      
      // Remove @ if present
      username = username.replace(/^@/, '');
      if (username) {
        usernames.push(username);
      }
    } catch (error) {
      console.error(`Failed to parse URL ${profileUrl}:`, error);
      continue;
    }
  }

  return usernames;
}

File: azure_functions/agents/Twitter_Apify_Py/supabase_retriever.py
import os
import json
from supabase import create_client

def get_supabase_client():
    """Create and return Supabase client"""
    URL = os.environ.get("SUPABASE_URL")
    KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")
    
    if not URL or not KEY:
        raise ValueError("SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY must be set")
    
    return create_client(URL, KEY)

def load_json_from_supabase(item_name: str):
    """
    Loads JSON content from a specific item in the 'items' table in Supabase.
    """
    try:
        supabase = get_supabase_client()
        res = supabase.table("items").select("payload").eq("name", item_name).single().execute()
        if res.data and "payload" in res.data:
            return res.data["payload"]
        else:
            return None
    except Exception as e:
        print(f"Error loading {item_name} from Supabase: {e}")
        return None

def get_monitored_twitter_usernames():
    """
    Get list of Twitter usernames to monitor from Supabase.
    Returns list of usernames (without @).
    """
    config = load_json_from_supabase("twitter_config")
    if not config:
        return []
    
    usernames = []
    monitored_urls = config.get("monitored_urls", [])
    
    for entry in monitored_urls:
        profile_url = entry.get("profile_url", "")
        if not profile_url:
            continue
        
        # Extract username from URL like "https://x.com/@Lica2216"
        # Handle both x.com and twitter.com
        try:
            # Get last part of URL
            username = profile_url.rstrip('/').split('/')[-1]
            # Remove @ if present
            username = username.lstrip('@')
            if username:
                usernames.append(username)
        except Exception as e:
            print(f"Failed to parse URL {profile_url}: {e}")
            continue
    
    return usernames


